File: Sudoku.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		fmt.Println(solve(scanner.Text()))
	}
}

func solve(input string) string {
	if s, ok := newSudoku().read(input).search(); ok {
		return s.String()
	} else {
		return "NO SOLUTION"
	}
}

type (
	View   int
	Coord  int
	Block  int
	State  int
	Count  int
	Sudoku struct {
		state [nCoord]State
		count [nBlock]Count
		queue []Coord
	}
)

const (
	unit   = 3
	size   = unit * unit
	nCoord = size * size * size
	nBlock = 4 * size * size
)

const (
	grd View = iota
	row
	col
	box
)

const (
	open State = iota
	fixed
	banned
)

const (
	done Count = size + 1
)

var (
	digits = []rune("123456789")

	parents  [nCoord][]Block
	children [nBlock][]Coord
)

func coord(i, j, k int) Coord {
	return Coord((i*size+j)*size + k)
}

func block(v View, p, q int) Block {
	return Block((int(v)*size+p)*size + q)
}

func init() {
	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			p := i/unit*unit + j/unit
			for k := 0; k < size; k++ {
				parents[coord(i, j, k)] = []Block{
					block(grd, i, j),
					block(row, i, k),
					block(col, j, k),
					block(box, p, k),
				}
			}
		}
	}
	for c, bs := range parents {
		for _, b := range bs {
			children[b] = append(children[b], Coord(c))
		}
	}
}

func newSudoku() *Sudoku {
	s := new(Sudoku)
	for c := range s.state {
		s.state[c] = open
	}
	for b := range s.count {
		s.count[b] = Count(size)
	}
	return s
}

func (s *Sudoku) clone() *Sudoku {
	ss := new(Sudoku)
	for c := range s.state {
		ss.state[c] = s.state[c]
	}
	for b := range s.count {
		ss.count[b] = s.count[b]
	}
	return ss
}

func (s *Sudoku) enqueue(c Coord) *Sudoku {
	s.queue = append(s.queue, c)
	return s
}

func (s *Sudoku) read(input string) *Sudoku {
	array := []rune(input)
	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			if i*size+j < len(array) {
				r := array[i*size+j]
				for k := 0; k < size; k++ {
					if r == digits[k] {
						s.enqueue(coord(i, j, k))
					}
				}
			}
		}
	}
	return s
}

func (s *Sudoku) String() string {
	result := make([]rune, size*size)
	for i := 0; i < size; i++ {
		for j := 0; j < size; j++ {
			result[i*size+j] = '.'
			for k := 0; k < size; k++ {
				if s.state[coord(i, j, k)] == fixed {
					result[i*size+j] = digits[k]
				}
			}
		}
	}
	return string(result)
}

func (s *Sudoku) search() (*Sudoku, bool) {

	for len(s.queue) > 0 {
		c := s.queue[0]
		s.queue = s.queue[1:]
		switch s.state[c] {
		case open:
			if ok := s.fix(c); !ok {
				return nil, false
			}
		case fixed:
			continue
		case banned:
			return nil, false
		}
	}

	min, bmin := done, -1
	for b, cnt := range s.count {
		if cnt < min {
			min = cnt
			bmin = b
		}
	}

	if min == done {
		return s, true
	}

	for _, c := range children[bmin] {
		if s.state[c] == open {
			if ss, ok := s.clone().enqueue(c).search(); ok {
				return ss, true
			}
		}
	}

	return nil, false
}

func (s *Sudoku) fix(c Coord) (ok bool) {
	defer func() {
		if err := recover(); err != nil {
			ok = false
		}
	}()
	s.state[c] = fixed
	for _, b := range parents[c] {
		s.markDone(b)
	}
	return true
}

func (s *Sudoku) markDone(b Block) {
	s.count[b] = done
	for _, c := range children[b] {
		if s.state[c] == open {
			s.ban(c)
		}
	}
}

func (s *Sudoku) ban(c Coord) {
	s.state[c] = banned
	for _, b := range parents[c] {
		if s.count[b] != done {
			s.countdown(b)
		}
	}
}

func (s *Sudoku) countdown(b Block) {
	s.count[b]--
	switch s.count[b] {
	case 0:
		panic("No choice!!")
	case 1:
		for _, c := range children[b] {
			if s.state[c] == open {
				s.enqueue(c)
			}
		}
	}
}

File: Sudoku.nim
from algorithm import fill
from queues import Queue, initQueue, enqueue, dequeue, len
from strutils import repeat

const
  unit = 3
  size = unit * unit
  done = size + 1
  digits = "123456789"

type
  View = enum grd, row, col, box
  Coord = range[0 .. size * size * size - 1]
  Group = range[0 .. 4 * size * size - 1]
  State = enum open, fixed, banned
  Count = range[0 .. done]
  Sudoku = ref object of RootObj
    state: array[Coord, State]
    count: array[Group, Count]
    queue: Queue[Coord]
  NoSolutionError = ref object of Exception

proc coord(i: int, j: int, k:int): Coord =
  (i * size + j) * size + k

proc group(v: View, p: int, q: int): Group =
  (v.ord * size + p) * size + q

var
  parents: array[Coord, seq[Group]]
  children: array[Group, seq[Coord]]

for i in 0 ..< size:
  for j in 0 ..< size:
    let p = i div unit * unit + j div unit
    for k in 0 ..< size:
      parents[coord(i, j, k)] = @[
        group(grd, i, j),
        group(row, i, k),
        group(col, j, k),
        group(box, p, k),
        ]

for g, _ in children:
  children[g] = @[]
for c, gs in parents:
  for g in gs:
    children[g].add(c)

proc newSudoku(): Sudoku =
  new result
  result.state.fill(open)
  result.count.fill(size)
  result.queue = initQueue[Coord]()

proc newSudoku(s: Sudoku): Sudoku =
  new result
  result.state = s.state
  result.count = s.count
  result.queue = initQueue[Coord]()

proc enqueue(s: Sudoku, c: Coord): Sudoku {.discardable.} =
  result = s
  result.queue.enqueue(c)

proc read(s: Sudoku, input: string): Sudoku =
  result = s
  for i in 0 ..< size:
    for j in 0 ..< size:
      if i * size + j < input.len:
        let k = digits.find(input[i * size + j])
        if k != -1:
          result.enqueue(coord(i, j, k))

proc `$`(s: Sudoku): string =
  result = repeat('.', size * size)
  for i in 0 ..< size:
    for j in 0 ..< size:
      for k in 0 ..< size:
        if s.state[coord(i, j, k)] == fixed:
          result[i * size + j] = digits[k]

proc fix(s: Sudoku, c: Coord)
proc mark(s: Sudoku, g: Group)
proc ban(s: Sudoku, c: Coord)
proc dec(s: Sudoku, g: Group)

proc search(s: Sudoku): Sudoku =

  while s.queue.len > 0:
    let c = s.queue.dequeue
    case s.state[c]
    of open: s.fix(c)
    of fixed: continue
    of banned: raise new NoSolutionError

  var
    min = done
    gmin: Group
  for g, cnt in s.count:
    if cnt < min:
      min = cnt
      gmin = g

  if min == done:
    return s

  for c in children[gmin]:
    if s.state[c] == open:
      try:
        return newSudoku(s).enqueue(c).search()
      except NoSolutionError:
        discard

  raise new NoSolutionError

proc fix(s: Sudoku, c: Coord) =
  s.state[c] = fixed
  for g in parents[c]:
    s.mark(g)

proc mark(s: Sudoku, g: Group) =
  s.count[g] = done
  for c in children[g]:
    if s.state[c] == open:
      s.ban(c)

proc ban(s: Sudoku, c: Coord) =
  s.state[c] = banned
  for g in parents[c]:
    if s.count[g] != done:
      s.dec(g)

proc dec(s: Sudoku, g: Group) =
  s.count[g].dec
  case s.count[g]
  of 0:
    raise new NoSolutionError
  of 1:
    for c in children[g]:
      if s.state[c] == open:
        s.enqueue(c)
  else:
    discard

proc solve(input: string): string =
  try:
    result = $newSudoku().read(input).search()
  except NoSolutionError:
    result = "NO SOLUTION"

try:
  while true:
    echo stdin.readLine().solve()
except IOError:
  discard
